### PR TITLE
docs: add concepts for subgraphs and multi-agent

### PR DESCRIPTION
**Pull request by vbarda**

### CODE REVIEW

File: docs/docs/concepts/low_level.md [hunk 1]
```diff
@@ -416,10 +417,109 @@ def my_node(state: State) -> State:
...
+def node(state: State):
+    # transform the state to the subgraph state
+    response = subgraph.invoke({"bar": state["foo"]})
+    # transform response back to the parent state
+    return {"foo": response["bar"]}
+
+builder = StateGraph(State)
+builder.add_node(node)
+...
+# note that instead of using the compiled subgraph we are using `node` function that is calling the subgraph
```
Comment (nfcampos) [inline docs/docs/concepts/low_level.md:515]: too long

File: docs/docs/concepts/low_level.md [hunk 2]
```diff
@@ -416,10 +417,109 @@ def my_node(state: State) -> State:
...
+class State(TypedDict):
+    foo: str
+
+class SubgraphState(TypedDict):
+    foo: str  # note that this key is shared with the parent graph state
+    bar: str
+
+# Define subgraph
+def subgraph_node(state: SubgraphState):
+    # note that this subgraph nodes can communicate with the parent graph via the shared "foo" key
```
Comment (nfcampos) [inline docs/docs/concepts/low_level.md:468]: typos

File: docs/docs/concepts/low_level.md [hunk 3]
```diff
@@ -416,10 +417,109 @@ def my_node(state: State) -> State:
...
+    return {"foo": state["foo"] + "bar"}
+
+subgraph_builder = StateGraph(SubgraphState)
+subgraph_builder.add_node(subgraph_node)
```
Comment (nfcampos) [inline docs/docs/concepts/low_level.md:472]: probably should add a note that extra keys in input/output will be ignored by root/subgraph

File: docs/docs/concepts/multi_agent.md [hunk 4]
```diff
@@ -1,138 +1,169 @@
...
 
-The primary benefits of this architecture are:
+* agent uses too many tools and makes poor decisions about which tool to call next
+* context grows too complex for a single agent to keep track of
+* there is a need for multiple specialization areas in the system (e.g. planner, researcher, math expert, etc.)
 
-* **Modularity**: Separate agents facilitate easier development, testing, and maintenance of agentic systems.
-* **Specialization**: You can create expert agents focused on specific domains, and compose them into more complex applications
-* **Control**: You can explicitly control how agents communicate (as opposed to relying on function calling)
+To tackle these, you might consider breaking your application into multiple smaller, independent agents and composing them into a **multi-agent system**. These independent agents can be as simple as a prompt and an LLM call, or as complex as a [ReAct](./agentic_concepts.md#react-implementation) agent.
```
Comment (nfcampos) [inline docs/docs/concepts/multi_agent.md:9]: Or even more complex...

File: docs/docs/concepts/multi_agent.md [hunk 5]
```diff
@@ -1,138 +1,169 @@
...
 
-## Multi-agent systems in LangGraph
+The primary benefits of using multi-agent systems are:
 
-### Agents as nodes
+* **Modularity**: Separate agents make it easier to develop, test, and maintain agentic systems.
```
Comment (nfcampos) [inline docs/docs/concepts/multi_agent.md:13]: I think this is a benefit of subgraphs, not multi agents, but tricky to make the distinction well

File: docs/docs/concepts/multi_agent.md [hunk 6]
```diff
@@ -1,138 +1,169 @@
...
-* Simple **LLM nodes**: single LLMs with custom prompts
-* **Subgraph nodes**: complex graphs called inside the orchestrator graph node
+![](./img/multi_agent/architectures.png)
 
-![](./img/multi_agent/subgraph.png)
+There are several ways to connect agents in a multi-agent system:
 
-### Agents as tools
+* **Network**: each agent can communicate with any other agent. Any agent can decide which other agent to call next.
+* **Supervisor**: each agent communicates with a single [supervisor](../tutorials/multi_agent/agent_supervisor/) agent. Supervisor agent makes decisions on which agent should be called next.
```
Comment (nfcampos) [inline docs/docs/concepts/multi_agent.md:24]: Worth clarifying whether multiple agents can be acting at the same time?

File: docs/docs/concepts/multi_agent.md [hunk 7]
```diff
@@ -1,138 +1,169 @@
...
+* **Custom multi-agent workflow**: each agent communicates with only a subset of agents. Parts of the flow are deterministic, and only some agents can decide which other agents to call next.
 
-Agents can also be defined as tools. In this case, the orchestrator agent (e.g. ReAct agent) would use a tool-calling LLM to decide which of the agent tools to call, as well as the arguments to pass to those agents.
+### Network
 
-You could also take a "mega-graph" approach – incorporating subordinate agents' nodes directly into the parent, orchestrator graph. However, this is not recommended for complex subordinate agents, as it would make the overall system harder to scale, maintain and debug – you should use subgraphs or tools in those cases.
+In this architecture, agents are defined as graph nodes. Each agent is connected to all other agents (many-to-many connections) and can decided which agent to call next. While very flexible, this architecture is inefficient and doesn't scale well as the number of agents grows. We recommend avoiding this architecture in production and using one of the below architectures instead.
```
Comment (nfcampos) [inline docs/docs/concepts/multi_agent.md:31]: Why inefficient? I think would be nice to give at least one example of why

File: docs/docs/concepts/multi_agent.md [hunk 8]
```diff
@@ -1,138 +1,169 @@
...
+* **Supervisor**: each agent communicates with a single [supervisor](../tutorials/multi_agent/agent_supervisor/) agent. Supervisor agent makes decisions on which agent should be called next.
+* **Supervisor (tool-calling)**: this is a special case of supervisor architecture. Individual agents can be represented as tools. In this case, a supervisor agent uses a tool-calling LLM to decide which of the agent tools to call, as well as the arguments to pass to those agents.
+* **Hierarchical**: you can define a multi-agent system with a supervisor of supervisors. This is a generalization of the supervisor architecture and allows for more complex control flows.
+* **Custom multi-agent workflow**: each agent communicates with only a subset of agents. Parts of the flow are deterministic, and only some agents can decide which other agents to call next.
 
-Agents can also be defined as tools. In this case, the orchestrator agent (e.g. ReAct agent) would use a tool-calling LLM to decide which of the agent tools to call, as well as the arguments to pass to those agents.
+### Network
```
Comment (nfcampos) [inline docs/docs/concepts/multi_agent.md:29]: Is this just a special case of custom workflow?

Comment (vbarda) [reply docs/docs/concepts/multi_agent.md:29]: yes, but by that logic every architecture is a special case of a custom workflow :)

File: docs/docs/concepts/low_level.md [hunk 9]
```diff
@@ -416,10 +417,112 @@ def my_node(state: State) -> State:
     return state
 ```
 
+## Subgraphs
+
+A subgraph is a [graph](#graphs) that is used as a [node](#nodes) in another graph. This is nothing more than the age-old concept of encapsulation, applied to LangGraph. Some reasons for using subgraphs are:
```
Comment (eyurtsev) [inline docs/docs/concepts/low_level.md:422]: Is it "used as a node" if it's being *invoked* from a node?

File: docs/docs/concepts/low_level.md [hunk 10]
```diff
@@ -416,10 +417,112 @@ def my_node(state: State) -> State:
...
+
+- building [multi-agent systems](./multi_agent.md)
+
+- when you want to reuse a set of nodes in multiple graphs, which maybe share some state, you can define them once in a subgraph and then use them in multiple parent graphs
```
Comment (eyurtsev) [inline docs/docs/concepts/low_level.md:426]: nit: set of nodes -> one or more nodes?

Comment (eyurtsev) [inline docs/docs/concepts/low_level.md:426]: nit: the `them` could become ambiguous due to both `nodes` and `graphs` being used as nouns

* you can define them once => you can define the nodes once

Comment (eyurtsev) [inline docs/docs/concepts/low_level.md:426]: nit: 
* which maybe share some state -> which may share some state

File: docs/docs/concepts/low_level.md [hunk 11]
```diff
@@ -416,10 +417,112 @@ def my_node(state: State) -> State:
...
+
+- when you want different teams to work on different parts of the graph independently, you can define each part as a subgraph, and as long as the subgraph interface (the input and output schemas) is respected, the parent graph can be built without knowing any details of the subgraph
```
Comment (eyurtsev) [inline docs/docs/concepts/low_level.md:428]: nit: these are long sentences for the bullet point list, perhaps consider capitalizing and using a `.` at the end of each item

File: docs/docs/concepts/low_level.md [hunk 12]
```diff
@@ -416,10 +417,112 @@ def my_node(state: State) -> State:
...
+
+There are two ways to add subgraphs to a parent graph:
+
+- add a node with the compiled subgraph: this is useful when the parent graph and the subgraph share state keys and you don't need to transform state on the way in or out
```
Comment (eyurtsev) [inline docs/docs/concepts/low_level.md:432]: * add a node with the compiled subgraph  => add the compiled subgraph as a node

File: docs/docs/concepts/low_level.md [hunk 13]
```diff
@@ -416,10 +417,112 @@ def my_node(state: State) -> State:
...
+
+```python
+builder.add_node("subgraph", subgraph_builder.compile())
+```
+
+- add a node with a function that invokes the subgraph: this is useful when the parent graph and the subgraph have different state schemas and you need to transform state before or after calling the subgraph
```
Comment (eyurtsev) [inline docs/docs/concepts/low_level.md:438]: * Invoke the subgraph directly from within a node

File: docs/docs/concepts/low_level.md [hunk 14]
```diff
@@ -416,10 +417,112 @@ def my_node(state: State) -> State:
...
+The simplest way to create subgraph nodes is by using a [compiled subgraph](#compiling-your-graph) directly. When doing so, it is **important** that the parent graph and the subgraph [state schemas](#state) share at least one key which they can use to communicate. If your graph and subgraph do not share any keys, you should use write a function [invoking the subgraph](#as-a-function) instead.
+
+!!! Note
+    If you pass extra keys to the subgraph node (i.e., in addition to the shared keys), they will be ignored by the subgraph node. Similarly, if you return extra keys from the subgraph, they will be ignored by the parent graph.
+
+```python
+from langgraph.graph import START, StateGraph
+from typing import TypedDict
+
+class State(TypedDict):
```
Comment (eyurtsev) [inline docs/docs/concepts/low_level.md:462]: IMO push `State` to the bottom -- after subgraph has been defined.

On the naming you could rename `foo` to `shared_foo` if you're worried about people seeing `foo` being shared.

**Review by eyurtsev — COMMENTED** [review]
just minor nits

**Pull request merged**